### PR TITLE
Add integration for Registration Forms fields

### LINKED ISSUE #120: Saving Registration form fields spinner appears on the wrong widget container
This is likely an issue since I swapped the vertical order of the widget columns to put the form fields on the left https://github.com/WPUserManager/wp-user-manager/commit/e6d276e328ca68b8be63d74e2cd4a69e72e97e7f

image

### PR DESCRIPTION
**Pull request by polevaultweb**
Also resolves #120

### CODE REVIEW

File: includes/wpum-fields/class-wpum-field.php [hunk 1]
```diff
@@ -121,6 +121,13 @@ class WPUM_Field {
 	 */
 	public $field_type;
 
+	/**
+	 * Check if the field is virtual
+	 * 
+	 * @var boolean
+	 */
+	protected $is_virtual = false;
```
Comment (polevaultweb) [inline includes/wpum-fields/class-wpum-field.php:129]: Is this needed anymore?

File: includes/wpum-fields/class-wpum-field.php [hunk 2]
```diff
@@ -143,6 +150,7 @@ public function __construct( $_id_or_field = false ) {
 
 		if ( $field ) {
 			$this->setup_field( $field );
+			do_action('wpum_field_'.$this->id.'_init', $this);
```
Comment (polevaultweb) [inline includes/wpum-fields/class-wpum-field.php:153]: Formatting needs fixing, and in other places

File: includes/wpum-fields/wpum-fields-functions.php [hunk 3]
```diff
@@ -632,3 +632,9 @@ function wpum_max_upload_size( $field_name = '', $custom_size = false ) {
 
 	return $output;
 }
+
+
+function wpum_get_field_class_name( $field ) {
+
+	return apply_filters( 'wpum_field_class', 'WPUM_Field_' . ucfirst( $field ), $field );
```
Comment (polevaultweb) [inline includes/wpum-fields/wpum-fields-functions.php:639]: Are we using this?

**Comment by polevaultweb** [comment]
@achchu93 there are a couple of other unresolved comments on this PR

File: includes/wpum-forms/class-wpum-form-registration.php [hunk 4]
```diff
@@ -532,7 +532,7 @@ public function done() {
 	 */
 	public function render_registration_form_fields($field, $key){
 		$registered_groups = array_column(wpum_get_registered_field_types(), 'fields');
-		$registered_fields = $registered_groups ? array_merge(...$registered_groups) : [];
+		$registered_fields = $registered_groups ? call_user_func_array('array_merge', $registered_groups) : [];
```
Comment (polevaultweb) [inline includes/wpum-forms/class-wpum-form-registration.php:535]: array_reduce($registered_groups, 'array_merge', []));

would this not be the < php 5.6 equivalent? https://www.php.net/manual/en/function.array-merge.php#121023

Comment (achchu93) [reply includes/wpum-forms/class-wpum-form-registration.php:535]: > array_reduce($registered_groups, 'array_merge', []));
> 
> would this not be the < php 5.6 equivalent? https://www.php.net/manual/en/function.array-merge.php#121023

Both works similarly. So i used this one. Workflows different for each one. So i dont think it is a big issue. If you still want it to be changed to `array_reduce` i can replace it

Comment (polevaultweb) [reply includes/wpum-forms/class-wpum-form-registration.php:535]: That's fine

**Pull request merged**